src/services/sessionService.ts

```diff
@@ -0,0 +1,184 @@
+    public readonly onDidChangeSession = this._onDidChangeSession.event;
+
+    private session: vscode.AuthenticationSession | undefined;
+    private loginPromise: Promise<void> | undefined;
+    private firstLoadPromise: Promise<void>;
+
+    private _publicApi: GitpodPublicApi | undefined;
+
+    get didFirstLoad() {
+        return this.firstLoadPromise;
+    }
+
+    constructor(
+        private readonly hostService: IHostService,
+        private readonly logger: ILogService
+    ) {
+        super();
+
+        this._register(vscode.authentication.onDidChangeSessions(e => this.handleOnDidChangeSessions(e)));
+        this.firstLoadPromise = this.tryLoadSession(false);
+    }
+
+    private async handleOnDidChangeSessions(e: vscode.AuthenticationSessionsChangeEvent) {
+        if (e.provider.id !== 'gitpod') {
+            return;
+        }
+        const oldSession = this.session;
+        this.session = undefined as vscode.AuthenticationSession | undefined;
+        await this.tryLoadSession(false);
+        let didChange = false;
+        if (oldSession && !this.session) {
+            // sign out
+            didChange = true;
+        } else if (!oldSession && this.session) {
+            // sign in
+            didChange = true;
+        } else if (oldSession?.id !== this.session?.id) {
+            // host changed
+            didChange = true;
+        }
+        if (didChange) {
+            this._publicApi?.dispose();
+            this._publicApi = undefined;
+            this._onDidChangeSession.fire();
+        }
+    }
+
+    isSignedIn() {
+        return !!this.session;
+    }
+
+    async signIn(gitpodHost?: string) {
+        if (this.loginPromise) {
+            this.logger.info(`Existing login in progress. Waiting for completion...`);
+            return this.loginPromise;
+        }
+
+        this.loginPromise = this.doSignIn(gitpodHost);
+        this.loginPromise.finally(() => this.loginPromise = undefined);
+        return this.loginPromise;
+    }
+
+    private async doSignIn(gitpodHost?: string) {
+        if (gitpodHost && new URL(this.hostService.gitpodHost).host !== new URL(gitpodHost).host) {
+            const changedSessionPromise = eventToPromise(this.onDidChangeSession);
+            const updated = await this.hostService.changeHost(gitpodHost);
+            if (!updated) {
+                return;
+            }
+            // wait until session get updated after host config changed
+            await changedSessionPromise;
+        }
+
+        if (this.isSignedIn()) {
+            return;
+        }
+
+        await this.tryLoadSession(true);
+
+        if (this.isSignedIn()) {
+            this.logger.info(`Successfully signed in`);
+        } else {
+            this.logger.error(`Failed to sign in`);
+        }
+    }
+
+    private async tryLoadSession(createIfNone: boolean) {
+        try {
+            if (this.session && !createIfNone) {
```

Comment:
```suggestion
            if (this.session || !createIfNone) {
```

Reply:
renamed to `force` :see_no_evil: , does that sounds better?